components/iam/pkg/oidc/service.go

```diff
@@ -97,43 +124,59 @@ func randString(size int) (string, error) {
 
 func (s *Service) GetClientConfigFromRequest(r *http.Request) (*ClientConfig, error) {
 	issuerParam := r.URL.Query().Get("issuer")
-	if issuerParam == "" {
-		return nil, errors.New("issuer param not specified")
+	stateParam := r.URL.Query().Get("state")
+	if issuerParam == "" && stateParam == "" {
+		return nil, fmt.Errorf("missing request parameters")
 	}
-	issuer, err := url.QueryUnescape(issuerParam)
-	if err != nil {
-		return nil, errors.New("bad issuer param")
+
+	if issuerParam != "" {
+		for _, value := range s.configsById {
+			if value.Issuer == issuerParam {
+				return value, nil
+			}
+		}
 	}
-	log.WithField("issuer", issuer).Trace("at GetClientConfigFromRequest")
 
-	for _, value := range s.configsById {
-		if value.Issuer == issuer {
-			return value, nil
+	if stateParam != "" {
+		state, err := decodeStateParam(stateParam)
+		if err != nil {
+			return nil, fmt.Errorf("bad state param")
+		}
+		config := s.configsById[state.ClientConfigID]
+		if config != nil {
+			return config, nil
 		}
 	}
-	return nil, errors.New("failed to find OIDC config for request")
+
+	return nil, fmt.Errorf("failed to find OIDC config for request")
+}
+
+type AuthenticateParams struct {
+	OAuth2Result     *OAuth2Result
+	Issuer           string
+	NonceCookieValue string
 }
 
-func (s *Service) Authenticate(ctx context.Context, oauth2Result *OAuth2Result, issuer string, nonceCookieValue string) (*AuthResult, error) {
-	rawIDToken, ok := oauth2Result.OAuth2Token.Extra("id_token").(string)
+func (s *Service) Authenticate(ctx context.Context, params AuthenticateParams) (*AuthFlowResult, error) {
+	rawIDToken, ok := params.OAuth2Result.OAuth2Token.Extra("id_token").(string)
 	if !ok {
-		return nil, errors.New("id_token not found")
+		return nil, fmt.Errorf("id_token not found")
 	}
 
-	verifier := s.verifierByIssuer[issuer]
+	verifier := s.verifierByIssuer[params.Issuer]
```

Comment:
Check the value exists with
```
verifier, ok := s.verifierByIssuer[...]
```

Reply:
FYI, map is not concurrency safe. So if something were to add to the map, or remove, you can get a panic here

Reply:
Thanks for the hint. 

I guess we'll apply a lot of flux in that area when reading from DB will be added, but it needs to be considered as well.